internal/component/kafka/consumer.go

```diff
@@ -35,28 +37,111 @@ type consumer struct {
+				}, func() {
+					consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
+				}); err != nil {
+					consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				}
+			} else {
+				err := consumer.doCallback(session, message)
+				if err != nil {
+					consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				}
+			}
+		}
+	}
+	if isBulkSubscribe {
 		if consumer.k.consumeRetryEnabled {
 			if err := retry.NotifyRecover(func() error {
-				return consumer.doCallback(session, message)
+				return consumer.doBulkCallback(session, messages, claim.Topic())
 			}, b, func(err error, d time.Duration) {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v. Retrying...", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Errorf("Error processing Kafka bulk messages: %s. Error: %v. Retrying...", claim.Topic(), err)
 			}, func() {
-				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
+				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s", claim.Topic())
 			}); err != nil {
-				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s. Error: %v.", claim.Topic(), err)
 			}
 		} else {
-			err := consumer.doCallback(session, message)
+			err := consumer.doBulkCallback(session, messages, claim.Topic())
 			if err != nil {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Errorf("Error processing Kafka message: %s. Error: %v.", claim.Topic(), err)
 			}
 		}
 	}
-
 	return nil
 }
 
+func (consumer *consumer) doBulkCallback(session sarama.ConsumerGroupSession, messages []*sarama.ConsumerMessage, topic string) error {
+	consumer.k.logger.Debugf("Processing Kafka bulk message: %s", topic)
+	handler, err := consumer.k.GetTopicBulkHandler(topic)
+	if err != nil {
+		return err
+	}
+	var messageValues []KafkaBulkMessageEntry
+	messageValues = make([]KafkaBulkMessageEntry, 0)
+
+	entryIds := GetEntryIds(len(messages))
+	for i, message := range messages {
+		if message != nil {
+			metadata := make(map[string]string)
```

Comment:
Please pre-alloc this too with a capacity of `len(message.Headers)`

Reply:
Done.